src/common/offline_transformations/src/pruning/propagate_masks.cpp

```diff
@@ -582,19 +591,93 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
     }
 };
 
+class ngraph::pass::mask_propagation::Reduce : public MatcherPass {
+public:
+    Reduce() {
+        auto inputs = pattern::any_input();
+        auto weights = pattern::any_input(pattern::has_static_shape());
+        auto pooling_by_reduce = pattern::wrap_type<opset6::ReduceMin, opset6::ReduceMax, opset6::ReduceMean>({inputs, weights});
+
+        ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+            const auto & pattern_map = m.get_pattern_value_map();
+            const auto m_weights = pattern_map.at(weights);
+            const auto & m_input = pattern_map.at(inputs);
+            const auto & m_output = pattern_map.at(pooling_by_reduce);
+
+            const auto constant = std::dynamic_pointer_cast<opset6::Constant>(m_weights.get_node_shared_ptr());
+            if (!constant) return false;
+
+            // Check reduce operation reduces only dimension without masks
+            if (auto input_mask = getMask(m_input)) {
+                auto output_mask = std::make_shared<Mask>(m_output.get_partial_shape().rank().get_length());
+                const auto reduce_dims = constant->cast_vector<int>();
+
+                auto input_mask_row = input_mask.get();
+                auto output_mask_row = output_mask.get();
+                input_mask->add_callback([output_mask_row](Mask::Ptr cur_mask) -> bool {
+                    cur_mask->copy_value_from_mask(output_mask_row);
+                    return true;
+                }, output_mask);
+                output_mask->add_callback([input_mask_row, reduce_dims](Mask::Ptr cur_mask) -> bool{
+                    // Propagate masks through dimension only if this dimension isn't reduced
+                    for (size_t dim = 0; dim < std::min(cur_mask->size(), input_mask_row->size()); ++dim)
+                        if (std::find(reduce_dims.begin(), reduce_dims.end(), dim) == reduce_dims.end())
+                            cur_mask->at(dim) = input_mask_row->at(dim);
+                        else if (cur_mask->at(dim) != input_mask_row->at(dim))
+                            cur_mask->initialize_dependencies();
```

Comment:
This is needed in cases when masks are propagating "down" through reduce layer. For example(PruneStopReduceLayerDown test case), convolution 28043 produces non empty mask, and on step of mul op processing, this zeros goes to convolution 28030, which unsuccessfully would try to set this zeros to reduce layer. But the problem is that as reduce layer mask didn't changed thus it don't notify and update their neighbors thus zero in convolution 6902 will be kept.

![image](https://user-images.githubusercontent.com/74656388/149347788-d45965eb-2697-4e48-b994-2241296aa841.png)